sql/core/src/main/scala/org/apache/spark/sql/execution/Exchange.scala

```diff
@@ -69,7 +72,16 @@ case class Exchange(newPartitioning: Partitioning, child: SparkPlan) extends Una
           }
         }
         val part = new HashPartitioner(numPartitions)
-        val shuffled = new ShuffledRDD[Row, Row, Row](rdd, part)
+        val shuffled = sort match {
+          case false => new ShuffledRDD[Row, Row, Row](rdd, part)
+          case true =>
+            val sortingExpressions = expressions.zipWithIndex.map {
+              case (exp, index) =>
+                new SortOrder(BoundReference(index, exp.dataType, exp.nullable), Ascending)
```

Comment:
I don't quite understand. If not using BoundReference here, we are likely to run into ArrayIndexOutOfBound if some columns get pruned or column order changed during optimization.